compiler/ir/stream/access_pattern.py

```diff
@@ -0,0 +1,242 @@
+            tuple(AffineConstantExpr(0) for _ in range(dim))
+            + tuple(AffineDimExpr(i) for i in range(self.num_dims - dim)),
+            [],
+            self.num_dims - dim,
+            0,
+        )
+        return type(self)(self.bounds[dim:], new_pattern)
+
+    def tile_dim(self, dim: int, template_bound: int) -> Self:
+        """
+        Returns a new access pattern with the `dim` dimension split up into two
+        This translates to creating two for loops with adjusted bounds from one for loop
+
+
+        For example:
+            (d0, d1, d2) -> d0 + d1 + d2
+        For `dim` = 1, `template_bound` = 2:
+            (d0, d1, d2, d3) -> d0 + 2 * d1 + d2 + d3
+
+        The bounds are split in similar fashion:
+        For example:
+            [2, 8, 2]
+        For `dim` = 1, `template_bound` = 2:
+            [2, 4, 2, 2]
+
+        """
+        transform_map = AffineMap(
+            num_dims=self.num_dims + 1,
+            num_symbols=0,
+            # (d0, d1, d2, ..., dim-1) -> (d0, d1, d2, ..., dim-1)
+            results=tuple(AffineDimExpr(i) for i in range(dim))
+            # (dim) -> (template_bound * dim + dim + 1)
+            + (AffineDimExpr(dim) * template_bound + AffineDimExpr(dim + 1),)
+            # (dim + 1, dim + 2, ...) -> (dim + 2, dim + 3, dim + 3)
+            + tuple(AffineDimExpr(i + 1) for i in range(dim + 1, self.num_dims)),
+        )
+        new_pattern = self.pattern.compose(transform_map)
+        bound_to_tile = self.bounds[dim]
+        tiled_bound = bound_to_tile // template_bound if bound_to_tile else None
+        new_bounds = (
+            self.bounds[:dim] + (tiled_bound, template_bound) + self.bounds[dim + 1 :]
+        )
+
+        return type(self)(new_bounds, new_pattern)
+
+
+@dataclass(frozen=True)
+class SchedulePattern(AccessPattern):
+    """
+    A schedule pattern is a pattern for a schedule of an operation.
+
+    Schedule patterns are constrained to have static bounds.
+    """
+
+    # constrain bounds to only be int
+    bounds: tuple[int, ...]
+
+    def __init__(self, bounds: Sequence[int], pattern: AffineMap):
+        if any(bound is None or bound <= 0 for bound in bounds):
+            raise ValueError(
+                "All bounds must be static, strictly positive integers for a schedule"
+            )
+
+        super().__init__(bounds, pattern)
+
+
+@dataclass(frozen=True)
+class TemplatePattern(AccessPattern):
+    """
+    Template pattern is a pattern for an accelerator template.
+
+    Templates should not be transformed through either tiling/rotating/others.
+    """
+
+    def __init__(self, bounds: Sequence[int], pattern: AffineMap):
+        super().__init__(bounds, pattern)
+
+    def tile_dim(self, dim: int, template_bound: int) -> Self:
+        raise RuntimeError("A template should not be tiled")
+
+    def rotate(self, dim: int) -> Self:
+        raise RuntimeError("A template should not be rotated")
```

Comment:
Instead of raising a RuntimeError, can we not just not implement these methods for this class, then Pyright will complain  whenever you want to call this?

Reply:
yes that's probably better, I will modify it to this behaviour!